src/caffe/layers/cudnn_conv_layer.cu

```diff
@@ -21,21 +21,54 @@ void CuDNNConvolutionLayer<Dtype>::Forward_gpu(
 
     // Forward through cuDNN in parallel over groups.
     for (int g = 0; g < this->group_; g++) {
+      Dtype alpha = 1.0;
+      Dtype beta = 0.0;
+
+      cudnnConvolutionFwdAlgo_t algo;
+
+      // get the desired convolution algorithm
+      CUDNN_CHECK(cudnnGetConvolutionForwardAlgorithm(handle_[g],
+        bottom_descs_[i],
+        filter_desc_,
+        conv_descs_[i],
+        top_descs_[i],
+        CUDNN_CONVOLUTION_FWD_NO_WORKSPACE,
+        0,  // memoryLimitInBytes,
+        &algo));
+
+      // get minimum size of the workspace needed for the desired algorithm
+      size_t workspaceSizeInBytes;
+
+      CUDNN_CHECK(cudnnGetConvolutionForwardWorkspaceSize(handle_[g],
+        bottom_descs_[i],
+        filter_desc_,
+        conv_descs_[i],
+        top_descs_[i],
+        algo,
+        &workspaceSizeInBytes));
+
+      void *workspace = NULL;
+
       // Filters.
       CUDNN_CHECK(cudnnConvolutionForward(handle_[g],
-          bottom_descs_[i], bottom_data + bottom_offset_ * g,
-          filter_desc_, weight + weight_offset_ * g,
-          conv_descs_[i],
-          top_descs_[i], top_data + top_offset_ * g,
-          CUDNN_RESULT_NO_ACCUMULATE));
+            reinterpret_cast<void *>(&alpha),
```

Comment:
Right, I'm not asking why the casts are there, but why the coefficients are being treated differently from the data. The code suggests that the coefficients are being passed untyped, while the data is typed, so that the data is safer than the coefficients. But from the headers it seems that everything is untyped.

Ideally one could a gain little safety with (templated, perhaps) wrappers, or perhaps y'all will introduce a typed interface at some point. But, to me it seems misleading to mark some arguments as unsafe when all are unsafe. Of course, marking every argument with `reinterpret_cast` would be noisy, so maybe it is better to leave them out and make a note of the unsafe interface...